waspc/cli/src/Wasp/Cli/Interactive.hs

```diff
@@ -0,0 +1,86 @@
+module Wasp.Cli.Interactive
+  ( askForInput,
+    askToChoose,
+    askForRequiredInput,
+  )
+where
+
+import Data.Foldable (find)
+import Data.List (intercalate)
+import qualified Data.Text as T
+import System.IO (hFlush, stdout)
+import Text.Read (readMaybe)
+import qualified Wasp.Util.Terminal as Term
+
+askForRequiredInput :: String -> IO String
+askForRequiredInput question = do
+  ensureNotNull . askForInput $ question
+
+askToChoose :: Show a => String -> [a] -> IO a
+askToChoose question options = case options of
+  [] -> error "Cannot ask to choose from empty list of options."
+  [singleOption] -> return singleOption
+  _multipleOptionsCase -> askUserToChoose
+  where
+    askUserToChoose = do
+      putStrLn $ Term.applyStyles [Term.Bold] question
+      putStrLn optionsToChooseFrom
+      answer <- prompt
+
+      handleOptionIndex answer >>= \case
+        Just option -> return option
+        Nothing -> handleOptionName answer
+
+    handleOptionIndex answer = do
+      let index = parseIndexOrDefault answer
+
+      case index of
+        Just i -> do
+          if i `elem` indexes
+            then return $ Just $ options !! (i - 1)
+            else Just <$> invalidOptionAction
+        Nothing -> return Nothing
+      where
+        -- The default option is the first one
+        parseIndexOrDefault input = if null input then Just 1 else readMaybe input :: Maybe Int
+
+    handleOptionName answer = do
+      let option = findElemByString answer
+      case option of
+        Just o -> return o
+        Nothing -> invalidOptionAction
+
+    invalidOptionAction = do
+      putStrLn $ Term.applyStyles [Term.Red] "Invalid selection, write the name or the index of the option."
+      askToChoose question options
+
+    findElemByString option = find (\o -> show o == option) options
+
+    optionsToChooseFrom = intercalate "\n" $ prependIndexToOption options
+    prependIndexToOption = zipWith (\i o -> showIndexWithStyle i ++ show o ++ noteForDefault i) indexes
+    noteForDefault i = if i == 1 then " (default)" else ""
+    showIndexWithStyle i = Term.applyStyles [Term.Yellow] $ "[" ++ show (i :: Int) ++ "] "
+    indexes = [1 .. length options]
```

Comment:
I ended up going to town with this one. I made it as "haskellish" as possible, check it out! It wasn't so bad before, but, now it should look more declarative, utilize maybes more, and functions are doing less, so all together should be simpler / nicer. It also uses a couple of a bit advanced haskell concepts (not too advanced, but not super beginner), so check those out also, they are useful to know and easy to learn.

```hs
askToChoose :: forall o. Show o => String -> NonEmpty o -> IO o
askToChoose _ (option :| []) = return option
askToChoose question options = do
  putStrLn $ Term.applyStyles [Term.Bold] question
  putStrLn showIndexedOptions
  answer <- prompt
  getOptionMatchingAnswer answer & maybe printErrorAndAskAgain return
  where
    getOptionMatchingAnswer :: String -> Maybe o
    getOptionMatchingAnswer "" = Just defaultOption
    getOptionMatchingAnswer answer =
      getOptionByIndex answer `mplus` getOptionByName answer

    getOptionByIndex :: String -> Maybe o
    getOptionByIndex idxStr =
      case readMaybe idxStr of
        Just idx | idx >= 1 && idx <= length options -> Just $ options NE.!! (idx - 1)
        _invalidIndex -> Nothing

    getOptionByName :: String -> Maybe o
    getOptionByName name = find ((== name) . show) options

    printErrorAndAskAgain = do
      putStrLn $ Term.applyStyles [Term.Red] "Invalid selection, write the name or the index of the option."
      askToChoose question options

    showIndexedOptions :: String
    showIndexedOptions =
      intercalate "\n" $ showIndexedOption <$> zip [1 ..] (NE.toList options)
      where
        showIndexedOption (idx, option) =
          showIndex idx <> " " <> show option <> (if isDefaultOption option then " (default)" else "")
        showIndex i = Term.applyStyles [Term.Yellow] $ "[" ++ show (i :: Int) ++ "]"

    defaultOption :: o
    defaultOption = NE.head options

    isDefaultOption :: o -> Bool
    isDefaultOption option = show option == show defaultOption
```

Reply:
I've used `<|>` instead of `mplus` since it makes the code more clear for me. Great changes :)